packages/core/lib/dlc/finance/CsoInfo.ts

```diff
@@ -0,0 +1,183 @@
+  const endOutcome = midPiece.endpoint;
+
+  const unit = eventDescriptor.unit;
+
+  const maxGain = Value.fromSats(
+    endOutcome * UNIT_MULTIPLIER[unit.toLowerCase()] - ONE_BTC_CONTRACT.sats,
+  );
+  const maxLoss = Value.fromSats(
+    ONE_BTC_CONTRACT.sats - startOutcome * UNIT_MULTIPLIER[unit.toLowerCase()],
+  );
+
+  const contractSize = Value.fromSats(contractInfo.totalCollateral);
+  const offerCollateral = contractSize.clone();
+  offerCollateral.sub(
+    Value.fromSats((maxGain.sats * contractSize.sats) / BigInt(1e8)),
+  );
+
+  return {
+    maxGain,
+    maxLoss,
+    minPayout,
+    maxPayout,
+    contractSize,
+    offerCollateral,
+  };
+};
+
+/**
+ * Get CsoInfo from OrderOffer or DlcOffer and validate
+ *
+ * @param {HasContractInfo & HasType} offer
+ * @returns {CsoInfo}
+ */
+export const getCsoInfoFromOffer = (
+  offer: HasContractInfo & HasType,
+): CsoInfo => {
+  if (
+    offer.type !== MessageType.DlcOfferV0 &&
+    offer.type !== MessageType.OrderOfferV0
+  )
+    throw Error('Only DlcOfferV0 and OrderOfferV0 currently supported');
+
+  return getCsoInfoFromContractInfo(offer.contractInfo);
+};
+
+/**
+ * Validate Payout Function for proper CSO format
+ *
+ * It should have 3 PayoutCurvePieces which consist of a flat line (maxLoss),
+ * ascending line (maxLoss to maxGain) and finally another flat line (maxGain)
+ *
+ * All PayoutCurvePieces should be type PolynomialPayoutCurvePieces
+ *
+ * @param {PayoutFunctionV0} payoutFunction
+ */
+export const validateCsoPayoutFunction = (
+  payoutFunction: PayoutFunctionV0,
+): void => {
+  assert(
+    payoutFunction.pieces.length === 3,
+    'CSO Payout Function must have 3 PayoutCurvePieces',
+  );
+  for (const [i, piece] of payoutFunction.pieces.entries()) {
+    assert(
+      piece.payoutCurvePiece.type === MessageType.PolynomialPayoutCurvePiece,
+      'CSO Payout Function PayoutCurvePieces must be PolynomialCurvePieces',
+    );
+
+    const payoutCurvePiece = piece.payoutCurvePiece as PolynomialPayoutCurvePiece;
+    const points = payoutCurvePiece.points;
+
+    // eventOutcome should always be ascending
+    assert(
+      points[0].eventOutcome < points[1].eventOutcome,
+      'CSO Payout Function PayoutCurvePiece point payout should be an ascending line',
+    );
+
+    // endpoints should always be ascending
+    let previousPiece;
+    if (i > 0) {
+      previousPiece = payoutFunction.pieces[i - 1];
+      assert(previousPiece.endpoint < piece.endpoint);
+    }
+
+    switch (i) {
+      case 0:
+        // endpoints should always be ascending
+        assert(payoutFunction.endpoint0 < piece.endpoint);
+
+        // maxLoss should be a flat line
+        assert(payoutFunction.endpointPayout0 === piece.endpointPayout);
+        assert(
+          points[0].outcomePayout === points[1].outcomePayout,
+          'CSO Payout Function maxLoss PayoutCurvePiece point should be a flat line',
+        );
+        break;
+      case 1:
```

Comment:
I believe this curve with still pass the validation. Is this a legal CSO curve?

![image](https://user-images.githubusercontent.com/12267041/173867046-8ba6b8bb-ea78-4d3c-a8e9-04faec39da56.png)

Reply:
Great catch guys, fixed in [2698d8c](https://github.com/AtomicFinance/node-dlc/pull/177/commits/2698d8c7fb2bc47de0527d4a537e10621eb71f22)